crates/ra_ide_api/src/completion/complete_dot.rs

```diff
@@ -28,13 +28,13 @@ fn complete_fields(acc: &mut Completions, ctx: &CompletionContext, receiver: Ty)
             Ty::Adt { def_id, .. } => {
                 match def_id.resolve(ctx.db)? {
                     Def::Struct(s) => {
-                        let variant_data = s.variant_data(ctx.db)?;
-                        for field in variant_data.fields() {
+                        for field in s.fields(ctx.db)? {
```

Comment:
I am not sure what's the best API here. The simplest thing would be to just expose `hir::TypeRef`, but that has two drawbacks:

* it only works for defs which originated in the source code
* it won't be able to account for type substitutions. 

So, we need to expose types somehow, and types should ideally have their generic parameters substituted, but def itself does not have substitutions either! But I think we could do 

```
Ty::Adt { def_id, subst } => {
....
    set_detail(field.ty(ctx.db)?.map(|ty| ty.subst(subst).to_string())))

```

Reply:
>Maybe Ty could have a method which returns the fields (if it is an ADT) with already substituted types. That could actually also be used in type inference itself, probably.

This probably means duplicating the `Def` API though :-( I don't know what is the best solution here, perhaps we need to introduce some generic `WithSubst<T>` wrapper?

Alternatively, we can merge def and type for adt, and treat a plain def as a type with empty substitution. I think this is the approach taken by Kotlin:

* here's a Def: a class: https://github.com/JetBrains/kotlin/blob/e640f01b4e630ba98b9d912c44e391fa9260233c/core/descriptors/src/org/jetbrains/kotlin/descriptors/ClassDescriptor.java

* you can get a new Def by calling substitute: https://github.com/JetBrains/kotlin/blob/e640f01b4e630ba98b9d912c44e391fa9260233c/core/descriptors/src/org/jetbrains/kotlin/descriptors/impl/AbstractClassDescriptor.java#L109

* it'll return a wrapper of the original Def which applies substitution on the fly: https://github.com/JetBrains/kotlin/blob/e640f01b4e630ba98b9d912c44e391fa9260233c/core/descriptors/src/org/jetbrains/kotlin/descriptors/impl/LazySubstitutingClassDescriptor.java